
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Release history &#8212; CPAT 3.0.0 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="#">CPAT 3.0.0 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Release history</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="toctree-wrapper compound">
</div>
<a class="reference internal image-reference" href="_images/cpat.png"><img alt="alternate text" src="_images/cpat.png" style="width: 325.0px; height: 100.0px;" /></a>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This documentation is for CPAT v3.0.0 or future versions. For documentation of CPAT v2.0.0
and older versions, please go to <a class="reference external" href="http://rna-cpat.sourceforge.net/">http://rna-cpat.sourceforge.net/</a></p>
</div>
<div class="section" id="release-history">
<h1>Release history<a class="headerlink" href="#release-history" title="Permalink to this headline">¶</a></h1>
<p><strong>CPAT v3.0.3</strong> (03/08/2021)</p>
<p>Update “cpat.py” to handle alternative start codens.</p>
<p><strong>CPAT v3.0.2</strong> (08/17/2020)</p>
<p>Update “make_logitModel.py” to make it compatible with “cpat.py”.</p>
<p><strong>CPAT v3.0.1</strong></p>
<p>Minor bug fixed regarding the output format.</p>
<p><strong>CPAT v3.0.0</strong></p>
<p>For many transcripts, the longest ORF may not be the real ORF. For example, in human genome,
the 2nd longest ORF of NM_198086 is the real ORF, and the 3rd longest ORF of NM_030915 is the
real ORF. Version 3.0.0 is released to address this problem.</p>
<ol class="arabic simple">
<li><p>If model is provided, CPAT can be used as an ORFfinder. It gives exactly the same results as <a class="reference external" href="https://www.ncbi.nlm.nih.gov/orffinder/">NCBI ORFfinder</a> does.</p></li>
<li><p>Search for all ORF candidates. The number of ORF reported is controlled by <code class="code docutils literal notranslate"><span class="pre">--min-orf</span></code> and <code class="code docutils literal notranslate"><span class="pre">--top-orf</span></code>.</p></li>
<li><p>In addition to basic ORF information (“ORF frame”, “ORF strand”, “ORF start”, “ORF end”, “ORF sequence”), it also reports “coding probability” for each ORF.</p></li>
<li><p>The best ORF will be selected (controlled by <code class="code docutils literal notranslate"><span class="pre">--best-orf</span></code>) either by <strong>ORF length</strong> or <strong>coding probability</strong>.</p></li>
</ol>
</div>
<div class="section" id="introduction">
<h1>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h1>
<p>CPAT is a bioinformatics tool to predict RNA’s coding probability based on the RNA sequence characteristics. To achieve this goal,  CPAT calculates scores of these 4 linguistic features
from a set of known protein-coding genes and another set of non-coding genes.</p>
<ol class="arabic simple">
<li><p>ORF size</p></li>
<li><p>ORF coverage</p></li>
<li><p><a class="reference external" href="http://nar.oxfordjournals.org/content/10/17/5303.abstract">Fickett TESTCODE</a></p></li>
<li><p><a class="reference external" href="http://nar.oxfordjournals.org/content/20/24/6441.abstract">Hexamer usage bias</a></p></li>
</ol>
<p>CPAT will then builds a <a class="reference external" href="https://en.wikipedia.org/wiki/Logistic_regression">logistic regression</a> model using these 4 features as predictor variables and the “protein-coding status” as
the response variable. After evaluating the performance and determining the probability cutoff, the model can be used to predict new RNA sequences.</p>
</div>
<div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="prerequisite">
<h2>Prerequisite<a class="headerlink" href="#prerequisite" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://www.python.org/download/releases/3.0/">python3.5</a> or later version</p></li>
<li><p><a class="reference external" href="http://numpy.scipy.org/">numpy</a></p></li>
<li><p><a class="reference external" href="http://www.r-project.org/">R</a></p></li>
</ol>
</div>
<div class="section" id="install-cpat-using-pip3">
<h2>install CPAT using pip3<a class="headerlink" href="#install-cpat-using-pip3" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ pip3 install CPAT
$ pip3 install CPAT --upgrade  # if you already have CPAT v2.0 installed
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>User need to download prebuilt <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/prebuilt_model/">logit model and hexamer table</a> for human, mouse, zebrafish and fly. For other species, we provide scripts to build these models (see below).</p></li>
</ul>
</div>
</div>
</div>
<div class="section" id="run-cpat-online">
<h1>Run <a class="reference external" href="https://wlcb.oit.uci.edu/cpat/">CPAT online</a><a class="headerlink" href="#run-cpat-online" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="https://wlcb.oit.uci.edu/cpat/">CPAT webserver</a></p>
<p><strong>Step1: Upload data to CPAT server. There are 3 different ways to uploada</strong></p>
<ul class="simple">
<li><p>Upload BED or FASTA format files from local disk. Files can be regular or compressed (<a href="#id1"><span class="problematic" id="id2">*</span></a>.gz, <a href="#id3"><span class="problematic" id="id4">*</span></a>.Z. <a href="#id5"><span class="problematic" id="id6">*</span></a>.z, <a href="#id7"><span class="problematic" id="id8">*</span></a>.bz, <a href="#id9"><span class="problematic" id="id10">*</span></a>.bz2, <a href="#id11"><span class="problematic" id="id12">*</span></a>.bzip2).</p></li>
<li><p>For small dataset, user can copy and paste data (in BED or FASTA format) directly to the text area.</p></li>
<li><p>For extremely larger dataset, user can save data in web server (http, https or ftp) first, then paste the data url to text area.</p></li>
</ul>
<p><strong>Step2: Select Select Species assembly</strong></p>
<p><strong>Step3: Click Submit button</strong></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>This web server only supports Human (hg19), Mouse (mm9 and mm10), Fly (dm3) and Zebrafish (Zv9).</p></li>
<li><p>When input file is BED format, the reference genome is required and the assembly version is important.</p></li>
</ul>
</div>
</div>
<div class="section" id="run-cpat-on-local-computer">
<h1>Run CPAT on local computer<a class="headerlink" href="#run-cpat-on-local-computer" title="Permalink to this headline">¶</a></h1>
<div class="section" id="input-files">
<h2>Input files<a class="headerlink" href="#input-files" title="Permalink to this headline">¶</a></h2>
<p>User need to provide a gene file (‘-g’), a logit model file (‘-d’), a hexamer frequency table file (‘-x’)
and specify the output file name(‘-o’). Gene file could be either in <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1.7">BED</a>.
or <a class="reference external" href="http://en.wikipedia.org/wiki/FASTA_format">FASTA</a> format. If in BED format, user also needs to specify reference genome sequence (‘-r’).</p>
<ul class="simple">
<li><p><a class="reference external" href="http://dldcc-web.brc.bcm.edu/lilab/liguow/CGI/cpat/dat/human_test.bed">BED</a> format file (regular text or compressed). <a class="reference external" href="http://dldcc-web.brc.bcm.edu/lilab/liguow/CGI/cpat/dat/human_test.bed">BED</a> file should be in standard 12-column format.</p></li>
<li><p><a class="reference external" href="http://dldcc-web.brc.bcm.edu/lilab/liguow/CGI/cpat/dat/human_test.mRNA.fa">FASTA</a> format file (regular text or compressed)</p></li>
<li><p>a URL pointing to data that are saved remotely (data could be either BED or FASTA, either regular text or compressed file). <a class="reference external" href="http://">http://</a>, <a class="reference external" href="https://">https://</a> and <a class="reference external" href="ftp://">ftp://</a> are supported.</p></li>
</ul>
</div>
<div class="section" id="command-line-options">
<h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
<dl>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">--version</span></kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
<dt><kbd><span class="option">-g <var>GENE_FILE</var></span>, <span class="option">--gene=<var>GENE_FILE</var></span></kbd></dt>
<dd><p>Genomic sequnence(s) of RNA in FASTA
(<a class="reference external" href="https://en.wikipedia.org/wiki/FASTA_format">https://en.wikipedia.org/wiki/FASTA_format</a>) or
standard 12-column BED
(<a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">https://genome.ucsc.edu/FAQ/FAQformat.html#format1</a>)
format. It is recommended to use <em>short</em> and <em>unique</em>
sequence identifiers (such as Ensembl transcript id)
in FASTA and BED file. If this is a BED file,
reference genome (‘-r/–ref’) should be specified. The
input FASTA or BED file could be a regular text file
or compressed file (<a href="#id16"><span class="problematic" id="id17">*</span></a>.gz, <a href="#id18"><span class="problematic" id="id19">*</span></a>.bz2) or accessible URL
(<a class="reference external" href="http://">http://</a>, <a class="reference external" href="https://">https://</a>, <a class="reference external" href="ftp://">ftp://</a>).</p>
</dd>
<dt><kbd><span class="option">-o <var>OUT_FILE</var></span>, <span class="option">--outfile=<var>OUT_FILE</var></span></kbd></dt>
<dd><p>The prefix of output files.</p>
</dd>
<dt><kbd><span class="option">-d <var>LOGIT_MODEL</var></span>, <span class="option">--logitModel=<var>LOGIT_MODEL</var></span></kbd></dt>
<dd><p>Logistic regression model. The prebuilt models for
Human, Mouse, Fly, Zebrafish are availablel. Run
‘make_logitModel.py’ to build logistic regression
model out of your own training datset.</p>
</dd>
<dt><kbd><span class="option">-x <var>HEXAMER_DAT</var></span>, <span class="option">--hex=<var>HEXAMER_DAT</var></span></kbd></dt>
<dd><p>The hexamer frequency table. The prebuilt tables for
Human, Mouse, Fly, Zebrafish are availablel. Run
‘make_hexamer_tab.py’ to make this table out of your
own training dataset.</p>
</dd>
<dt><kbd><span class="option">-r <var>REF_GENOME</var></span>, <span class="option">--ref=<var>REF_GENOME</var></span></kbd></dt>
<dd><p>Reference genome sequences in FASTA format. Reference
genome file will be indexed automatically (produce
<a href="#id20"><span class="problematic" id="id21">*</span></a>.fai file along with the original <a href="#id22"><span class="problematic" id="id23">*</span></a>.fa file within
the same directory) if hasn’t been done. Ignore this
option if FASTA file was provided to ‘-g/–gene’.</p>
</dd>
<dt><kbd><span class="option">--antisense</span></kbd></dt>
<dd><p>Also search for ORFs from the anti-sense strand.
<em>Sense strand</em> (or coding strand) is DNA strand that
carries the translatable code in the 5′ to 3′
direction. default=False (i.e. only search for ORFs
from the sense strand)</p>
</dd>
<dt><kbd><span class="option">--start=<var>START_CODONS</var></span></kbd></dt>
<dd><p>Start codon used by ORFs. Use ‘T’ instead of ‘U’.
default=ATG</p>
</dd>
<dt><kbd><span class="option">--stop=<var>STOP_CODONS</var></span></kbd></dt>
<dd><p>Stop codons used by ORFs. Multiple stop codons should
be separated by ‘,’. Use ‘T’ instead of ‘U’.
default=TAG,TAA,TGA</p>
</dd>
<dt><kbd><span class="option">--min-orf=<var>MIN_ORF_LEN</var></span></kbd></dt>
<dd><p>Minimum ORF length in nucleotides.  default=75</p>
</dd>
<dt><kbd><span class="option">--top-orf=<var>N_TOP_ORF</var></span></kbd></dt>
<dd><p>Number of ORF candidates. Many RNAs have dozens of
possible ORFs, in most cases, the real ORF is ranked
(by size) in the top several. To increase speed, we do
not need to calculate “Fickett score”, “Hexamer score”
and “coding probability” for all of them. default=5</p>
</dd>
<dt><kbd><span class="option">--width=<var>LINE_WIDTH</var></span></kbd></dt>
<dd><p>Line width of output ORFs in FASTA format.
default=100</p>
</dd>
<dt><kbd><span class="option">--log-file=<var>LOG_FILE</var></span></kbd></dt>
<dd><p>Name of log file. default=”CPAT_run_info.log”</p>
</dd>
<dt><kbd><span class="option">--best-orf=<var>MODE</var></span></kbd></dt>
<dd><p>Criteria to select the best ORF: “l”=length, selection
according to the “ORF length”; “p”=probability,
selection according to the “coding probability”.
default=”p”</p>
</dd>
<dt><kbd><span class="option">--verbose</span></kbd></dt>
<dd><p>Logical to determine if detailed running information
is printed to screen.</p>
</dd>
</dl>
</dd>
</dl>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<p>Use FASTA file as input:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cpat.py -x Human_Hexamer.tsv --antisense -d Human_logitModel.RData --top-orf=5 -g Human_test_coding_mRNA.fa -o output1
</pre></div>
</div>
<p>Use BED file as input. ‘-r’ is required:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cpat.py -x Human_Hexamer.tsv --antisense -d Human_logitModel.RData --top-orf=5 -g Human_test_coding_mRNA_hg19.bed -r hg19.fa -o output2
</pre></div>
</div>
</div>
<div class="section" id="output-files">
<h2>output files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>output.ORF_seqs.fa: The top ORF sequences (at least 75 nucleotides long) in FASTA format.</p></li>
<li><p>output.ORF_prob.tsv: ORF information (strand, frame, start, end, size, Fickett TESTCODE score, Hexamer score) and coding probability)</p></li>
<li><p>output.ORF_prob.best.tsv: The information of the best ORF. This file is a subset of “output.ORF_prob.tsv”</p></li>
<li><p>output.no_ORF.txt: Sequence IDs or BED entries with no ORF found. Should be considered as non-coding.</p></li>
<li><p>output.r: Rscript file.</p></li>
<li><p>CPAT_run_info.log: log file</p></li>
</ul>
</div>
</div>
<div class="section" id="build-hexamer-table">
<h1>Build hexamer table<a class="headerlink" href="#build-hexamer-table" title="Permalink to this headline">¶</a></h1>
<p><strong>make_hexamer_tab.py</strong> calculates the in frame hexamer (6mer) frequency from CDS sequence in fasta format. The CDS is mRNA sequence that removes UTR.
This table is required by CPAT to calculate the hexamer usage score. Users can download prebuilt hexamer tables (Human, Mouse, Fly, Zebrafish) from <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/prebuilt_model/">here</a></p>
<dl>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">--version</span></kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
<dt><kbd><span class="option">-c <var>CODING_FILE</var></span>, <span class="option">--cod=<var>CODING_FILE</var></span></kbd></dt>
<dd><p>Coding sequence (must be CDS without UTR, i.e. from start coden to stop coden) in fasta format.
User can get CDS sequence of a bed file using <a class="reference external" href="http://genome.ucsc.edu/cgi-bin/hgTables?org=Human&amp;db=hg19&amp;hgsid=289407045&amp;hgta_doMainPage=1">UCSC table browser</a></p>
</dd>
<dt><kbd><span class="option">-n <var>NONCODING_FILE</var></span>, <span class="option">--noncod=<var>NONCODING_FILE</var></span></kbd></dt>
<dd><p>Noncoding sequences in fasta format</p>
</dd>
</dl>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ make_hexamer_tab.py -c Human_coding_transcripts_CDS.fa   -n Human_noncoding_transcripts_RNA.fa &gt;Human_Hexamer.tsv

$ head Human_Hexamer.tsv

hexamer        coding  noncoding
GAACGT 0.000114999540425       6.20287252729e-05
CTTCTT 0.000280298143192       0.000464526231488
CACCCT 0.000254883880114       0.000337895737524
GAACGG 0.000178535198119       5.8077265737e-05
GAACGC 0.000136389878516       6.03746259323e-05
GAACGA 0.00015830968042        5.87205265917e-05
CACCCA 0.000258696019576       0.000448628498937
CTTCTA 0.000147508618612       0.000280645521457
CACCCC 0.000328479350276       0.000342582352322
...
</pre></div>
</div>
</div>
<div class="section" id="build-logit-model">
<h1>Build Logit model<a class="headerlink" href="#build-logit-model" title="Permalink to this headline">¶</a></h1>
<p>Build logistic regression model (“prefix.logit.RData”) required by CPAT. This program will output
3 files:</p>
<ul class="simple">
<li><p>prefix.feature.xls: A table contains features calculated from training datasets (coding and noncoding gene lists).</p></li>
<li><p>prefix.logit.RData: logit model required by CPAT (if R was installed).</p></li>
<li><p>prefix.make_logitModel.r: R script to build the above logit model.</p></li>
</ul>
<p>Note: Users can download prebuilt logit models (Human, Mouse, Fly, Zebrafish) from <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/prebuilt_model">here</a></p>
<dl>
<dt>Options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">--version</span></kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
<dt><kbd><span class="option">-c <var>CODING_FILE</var></span>, <span class="option">--cgene=<var>CODING_FILE</var></span></kbd></dt>
<dd><p>Genomic sequnences of protein-coding RNAs in FASTA
(<a class="reference external" href="https://en.wikipedia.org/wiki/FASTA_format">https://en.wikipedia.org/wiki/FASTA_format</a>) or
standard 12-column BED
(<a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">https://genome.ucsc.edu/FAQ/FAQformat.html#format1</a>)
format. It is recommended to use <em>short</em> and <em>unique</em>
sequence identifiers (such as Ensembl transcript id)
in FASTA and BED file. The input FASTA or BED file
could be a regular text file or compressed file (<a href="#id25"><span class="problematic" id="id26">*</span></a>.gz,
<a href="#id27"><span class="problematic" id="id28">*</span></a>.bz2) or accessible URL (<a class="reference external" href="http://">http://</a>, <a class="reference external" href="https://">https://</a>, <a class="reference external" href="ftp://">ftp://</a>).
When BED file is provided, use the ORF defined in the
BED file (the 7th and 8th columns in BED file define
the positions of ‘start codon, and ‘stop codon’,
respectively). When FASTA file is provided, searching
for the longet ORF. For well annotated genome, we
recommend using BED file as input because the longest
ORF predicted from RNA sequence might not be the real
ORF. If this is a BED file, reference genome
(‘-r/–ref’) should be specified.</p>
</dd>
<dt><kbd><span class="option">-n <var>NONCODING_FILE</var></span>, <span class="option">--ngene=<var>NONCODING_FILE</var></span></kbd></dt>
<dd><p>Genomic sequences of non-coding RNAs in FASTA
(<a class="reference external" href="https://en.wikipedia.org/wiki/FASTA_format">https://en.wikipedia.org/wiki/FASTA_format</a>) or
standard 12-column BED
(<a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">https://genome.ucsc.edu/FAQ/FAQformat.html#format1</a>)
format. It is recommended to use <em>short</em> and <em>unique</em>
sequence identifiers (such as Ensembl transcript id)
in FASTA and BED file. The input FASTA or BED file
could be a regular text file or compressed file (<a href="#id29"><span class="problematic" id="id30">*</span></a>.gz,
<a href="#id31"><span class="problematic" id="id32">*</span></a>.bz2) or accessible URL (<a class="reference external" href="http://">http://</a>, <a class="reference external" href="https://">https://</a>, <a class="reference external" href="ftp://">ftp://</a>).
If this is a BED file, reference genome (‘-r/–ref’)
should be specified.</p>
</dd>
<dt><kbd><span class="option">-o <var>OUT_FILE</var></span>, <span class="option">--outfile=<var>OUT_FILE</var></span></kbd></dt>
<dd><p>The prefix of output files.</p>
</dd>
<dt><kbd><span class="option">-x <var>HEXAMER_DAT</var></span>, <span class="option">--hex=<var>HEXAMER_DAT</var></span></kbd></dt>
<dd><p>Hexamer frequency table. CPAT has prebuilt hexamer
frequency tables for Human, Mouse, Fly, Zebrafish. Run
‘make_hexamer_tab.py’ to generate this table.</p>
</dd>
<dt><kbd><span class="option">-r <var>REF_GENOME</var></span>, <span class="option">--ref=<var>REF_GENOME</var></span></kbd></dt>
<dd><p>Reference genome sequences in FASTA format. Ignore
this option if mRNA sequences file was provided to
‘-g’. Reference genome file will be indexed
automatically if the index file  <a href="#id33"><span class="problematic" id="id34">*</span></a>.fai) does not
exist.</p>
</dd>
<dt><kbd><span class="option">-s <var>START_CODONS</var></span>, <span class="option">--start=<var>START_CODONS</var></span></kbd></dt>
<dd><p>Start codon (use ‘T’ instead of ‘U’) used to define
the start of open reading frame (ORF). default=ATG</p>
</dd>
<dt><kbd><span class="option">-t <var>STOP_CODONS</var></span>, <span class="option">--stop=<var>STOP_CODONS</var></span></kbd></dt>
<dd><p>Stop codon (use ‘T’ instead of ‘U’) used to define the
end of open reading frame (ORF). Multiple stop codons
are separated by ‘,’. default=TAG,TAA,TGA</p>
</dd>
<dt><kbd><span class="option">--min-orf=<var>MIN_ORF_LEN</var></span></kbd></dt>
<dd><p>Minimum ORF length in nucleotides.  default=30</p>
</dd>
<dt><kbd><span class="option">--log-file=<var>LOG_FILE</var></span></kbd></dt>
<dd><p>Name of log file.
default=”make_logitModel_run_info.log”</p>
</dd>
<dt><kbd><span class="option">--verbose</span></kbd></dt>
<dd><p>Logical to determine if detailed running information
is printed to screen.</p>
</dd>
</dl>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ make_logitModel.py  -x Human_Hexamer.tsv -c Human_coding_transcripts_mRNA.fa -n Human_noncoding_transcripts_RNA.fa -o Human

Process protein coding transcripts: Human_coding_transcripts_mRNA.fa
Input gene file is in FASTA format
Process non coding transcripts: Human_noncoding_transcripts_RNA.fa
Input gene file is in FASTA format
build logi model ...
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred

#or use BED file as input
$ make_logitModel.py  -x Human_Hexamer.tsv -c Human_coding_transcripts_hg19.bed -n Human_noncoding_transcripts_hg19.bed  -r /database/hg19.fa  -o Human
</pre></div>
</div>
</div>
<div class="section" id="use-cpat-to-detect-orf">
<h1>Use CPAT to detect ORF<a class="headerlink" href="#use-cpat-to-detect-orf" title="Permalink to this headline">¶</a></h1>
<p>It is perfectly fine to use CPAT to find ORFs. And CPAT will gives exactly the same results as <a class="reference external" href="https://www.ncbi.nlm.nih.gov/orffinder/">NCBI ORFfinder</a></p>
<div class="section" id="prepare-data">
<h2>Prepare data<a class="headerlink" href="#prepare-data" title="Permalink to this headline">¶</a></h2>
<p>Below is the mRNA sequence of protein-coding gene <a class="reference external" href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=UQCR10">UQCR10</a>. Copy and save it as “test.fa”.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">NM_013387</span><span class="o">.</span><span class="mi">4</span>
<span class="n">GCGGTGGCGCGAGTTGGACTGTGAAGAAACATGGCGGCCGCGACGTTGAC</span>
<span class="n">TTCGAAATTGTACTCCCTGCTGTTCCGCAGGACCTCCACCTTCGCCCTCA</span>
<span class="n">CCATCATCGTGGGCGTCATGTTCTTCGAGCGCGCCTTCGATCAAGGCGCG</span>
<span class="n">GACGCTATCTACGACCACATCAACGAGGGGAAGCTGTGGAAACACATCAA</span>
<span class="n">GCACAAGTATGAGAACAAGTAGTTCCTTGGAGGCCCCCATCCAGGCCAGA</span>
<span class="n">AGGACCAGGTCCACCCAGCAGCTGTTTGCCCAGAGCTGGAGCCTCAGCTT</span>
<span class="n">GAAGATGATGCTCAAGGTACTCTTCATGGACCACCATTCGCTGTTGGCAA</span>
<span class="n">GAAACGGCTTTACTTACAAAACAGACTCTTTACCTTCTGCTGTGTTTGAA</span>
<span class="n">GTATGTTTAGTCAGCATGCTCAGGAAATAAATGTGAATTGCCCTTGAGAC</span>
<span class="n">CTGCTTCTACATTGGTTGCTTTGTTAACTCTACCTGATCTTCACTTGTCA</span>
<span class="n">GTAATTTGAGACCACTTCAAAGCCCTCTGCAAACACCCCAAAGGCAGAAT</span>
<span class="n">CTGCTATTTTGAGTTTTCCATTAACTTCCAAAGAATTCTGGTTTTCAAAA</span>
<span class="n">CAGGAGCCAGAGTTGGAGATATTACAGTCAACTTTGGCTTCTAAGCCAGT</span>
<span class="n">AATTCCATTCTTAAATACCTCACTGTCTTGGCCATGGGGAAGCACTATGG</span>
<span class="n">CCTCAGCTGGGGGAAAGACCCTGGCCTAGGGGTCTTAGCCACTCCCCACC</span>
<span class="n">CTAGGGTATAGTTCAGGGGTATCCAATCCTTTGGCTTCCCTGGGCCATGT</span>
<span class="n">TGGAAGAATTGTCTTGGGCCACACATAAAATACAGTAACCATAGCTGATG</span>
<span class="n">AGCTAAAACAAAAAACAATGGTTTGTGCAAAAATCTCATAATGTTTTAAT</span>
<span class="n">AAAGTTGAAGAATTTG</span>
</pre></div>
</div>
</div>
<div class="section" id="run-cpat">
<h2>Run CPAT<a class="headerlink" href="#run-cpat" title="Permalink to this headline">¶</a></h2>
<p>The command to run cpat.py is as below</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cpat.py -x Human_Hexamer.tsv  -d  Human_logitModel.RData  --top-orf=100  --antisense -g test.fa -o output
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>You must specify <code class="code docutils literal notranslate"><span class="pre">--antisense</span></code>, otherwise, it will only search ORFs from the sense strand.</p></li>
<li><p>You also specify <code class="code docutils literal notranslate"><span class="pre">--top-orf</span></code> to a big number to report all the ORFs.</p></li>
<li><p>The <code class="code docutils literal notranslate"><span class="pre">--min-orf</span></code> is set to 75 by default, same as <a class="reference external" href="https://www.ncbi.nlm.nih.gov/orffinder/">NCBI ORFfinder</a>.</p></li>
</ul>
</div>
</div>
<div class="section" id="check-the-results">
<h2>Check the results<a class="headerlink" href="#check-the-results" title="Permalink to this headline">¶</a></h2>
<p>A total of 8 ORFs were found (sorted by the ORF size, the 7th column). If you copy and paste the same sequence
to <a class="reference external" href="https://www.ncbi.nlm.nih.gov/orffinder/">NCBI ORFfinder</a> web server, you will get <strong>exactly the same</strong> results.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$more output.ORF_prob.tsv
ID      mRNA    ORF_strand      ORF_frame       ORF_start       ORF_end ORF     Fickett Hexamer Coding_prob
NM_013387.4_ORF_1       916     -       2       327     1       327     1.103   0.28998918917275        0.792763525921043
NM_013387.4_ORF_2       916     +       2       209     430     222     1.1605  0.0674464550896935      0.271842476390681
NM_013387.4_ORF_3       916     -       1       889     695     195     0.9192  -0.32000518247443       0.0113140534730678
NM_013387.4_ORF_4       916     +       1       31      222     192     1.2952  0.600469985268255       0.915129459422605
NM_013387.4_ORF_5       916     -       1       337     197     141     1.1626  0.133867810597757       0.185245402415541
NM_013387.4_ORF_6       916     -       3       119     3       117     1.2673  0.442351820001225       0.618496534888714
NM_013387.4_ORF_7       916     -       3       842     735     108     0.5832  -0.19401829042094       0.00290794398512764
NM_013387.4_ORF_8       916     +       3       684     761     78      0.7415  -0.154613060436537      0.00454929869181486
</pre></div>
</div>
<p>CPAT also provides <em>Fickett’s TESTCODE score</em>, <em>Hexamer score</em> and <em>coding probability</em> for each ORF, to help you determine which one is more likely the <em>real</em> ORF.
For most mRNAs, the largest ORF is also the most likely one, but not always. In this particular example, ORF_4 is the most likely one to code for protein since it has
the highest coding probability. This can be demonstrated by BLATing the 8 ORF sequences to the reference genome.</p>
<div class="figure align-default">
<a class="reference internal image-reference" href="_images/UQCR10.jpg"><img alt="_images/UQCR10.jpg" src="_images/UQCR10.jpg" style="width: 500.0px; height: 90.0px;" /></a>
</div>
</div>
</div>
<div class="section" id="how-to-choose-cutoff">
<h1>How to choose cutoff<a class="headerlink" href="#how-to-choose-cutoff" title="Permalink to this headline">¶</a></h1>
<p><strong>Optimum cutoff were determined from TG-ROC</strong>.</p>
<ul class="simple">
<li><p>Human coding probability (CP) cutoff: 0.364 (CP &gt;=0.364 indicates coding sequence, CP &lt; 0.364 indicates noncoding sequence) (see performance figure D)</p></li>
<li><p>Mouse coding probability (CP) cutoff: 0.44</p></li>
<li><p>Fly coding probability (CP) cutoff: 0.39</p></li>
<li><p>Zebrafish coding probability (CP) cutoff: 0.38</p></li>
</ul>
<p><strong>Here we provide the R code and the data that we used to generate</strong> <a class="reference external" href="http://nar.oxfordjournals.org/content/41/6/e74/F3.expansion.html">Figure 3</a> <strong>in our paper</strong>. <strong>Note the</strong> <a class="reference external" href="http://cran.r-project.org/web/packages/ROCR/index.html">ROCR</a> <strong>library is required to run our R code</strong>.</p>
<p>1) Download R code and data from <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/Figure3_data/">here</a></p>
<p>2) Put the R code and the data table in the same folder</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ls
10Fold_CrossValidation.r       Human_train.dat
</pre></div>
</div>
<p>3) Run the R code from command line or console. The R code will perform 10-fold cross validation and generate Figure_3.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ Rscript 10Fold_CrossValidation.r     # install ROCR before running this code

Loading required package: gplots
Attaching package: ‘gplots’
The following object is masked from ‘package:stats’:

    lowess

Loading required package: methods
Warning message:
package ‘gplots’ was built under R version 3.1.2
[1] &quot;ID&quot;      &quot;mRNA&quot;    &quot;ORF&quot;     &quot;Fickett&quot; &quot;Hexamer&quot; &quot;Label&quot;
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
Warning message:
glm.fit: fitted probabilities numerically 0 or 1 occurred
null device
          1
</pre></div>
</div>
</div>
<div class="section" id="how-to-prepare-training-dataset">
<h1>How to prepare training dataset<a class="headerlink" href="#how-to-prepare-training-dataset" title="Permalink to this headline">¶</a></h1>
<p>We prebuild hexamer tables and logit models for <a class="reference external" href="https://sourceforge.net/projects/rna-cpat/files/v1.2.2/prebuilt_model/">human, mouse, fly and zebrafish</a>.
If you want to run CPAT for other species, you need to prepare your own training data. These two files are required when you run <em>make_hexamer_tab.py</em> and <em>make_logitModel.py</em>.</p>
<ul class="simple">
<li><p>It’s better to have balanced training dataset (i.e. the number of coding sequences is roughly equal to the number of noncoding sequences).</p></li>
<li><p>If the genome of the species you are working on is NOT well annotated and does not have enough “coding” and “noncoding” genes to build the training data, you could build your model using data from other species that is evolutionary close to the species you are working on.</p></li>
</ul>
</div>
<div class="section" id="evaluating-performance">
<h1>Evaluating Performance<a class="headerlink" href="#evaluating-performance" title="Permalink to this headline">¶</a></h1>
<p>Combinatorial effects of 3 major features. 10,000 coding genes (red dots) and 10,000 noncoding genes (blue dots) are clearly separated into two clusters. (below figure)</p>
<div class="figure align-default">
<a class="reference internal image-reference" href="_images/Figure_1A_features_3D.png"><img alt="_images/Figure_1A_features_3D.png" src="_images/Figure_1A_features_3D.png" style="width: 600.0px; height: 600.0px;" /></a>
</div>
<p>Performance evaluation using 10-fold cross validation (10,000 coding genes and 10,000 noncoding genes). Blue dotted curves represent the
10-fold cross validations, red solid curve represents the averaged curve between 10 runs of
validations. (A) ROC curve. (B) PR (precision-recall) curve. (C) Accuracy vs cutoff value.
(D) Two graphic ROC curve to determine the optimum cutoff value.</p>
<div class="figure align-default">
<a class="reference internal image-reference" href="_images/CPAT_performance.png"><img alt="_images/CPAT_performance.png" src="_images/CPAT_performance.png" style="width: 600.0px; height: 600.0px;" /></a>
</div>
</div>
<div class="section" id="comparison">
<h1>Comparison<a class="headerlink" href="#comparison" title="Permalink to this headline">¶</a></h1>
<p>To compare CPAT with CPC and PhyloCSF, we build an independent testing dataset that composed
of 4,000 high quality protein coding genes from Refseq annotation and 4,000 lincRNAs from
Human lincRNA catalog (Cabili et al., 2011). All 8000 genes were not included in the training
dataset of CPAT.</p>
<div class="figure align-default">
<a class="reference internal image-reference" href="_images/Figure_4.png"><img alt="_images/Figure_4.png" src="_images/Figure_4.png" style="width: 600.0px; height: 600.0px;" /></a>
</div>
<div class="figure align-default">
<a class="reference internal image-reference" href="_images/Figure_S2.png"><img alt="_images/Figure_S2.png" src="_images/Figure_S2.png" style="width: 600.0px; height: 600.0px;" /></a>
</div>
</div>
<div class="section" id="license">
<h1>LICENSE<a class="headerlink" href="#license" title="Permalink to this headline">¶</a></h1>
<p>CPAT is distributed under <a class="reference external" href="http://www.gnu.org/copyleft/gpl.html">GNU General Public License</a></p>
<p>This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version. This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details. You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301 USA</p>
</div>
<div class="section" id="reference">
<h1>Reference<a class="headerlink" href="#reference" title="Permalink to this headline">¶</a></h1>
<p>Wang, L., Park, H. J., Dasari, S., Wang, S., Kocher, J.-P., &amp; Li, W. (2013). CPAT: Coding-Potential Assessment Tool using an alignment-free logistic regression model. Nucleic Acids Research, 41(6), e74. <a class="reference external" href="http://nar.oxfordjournals.org/content/41/6/e74">doi:10.1093/nar/gkt006</a></p>
</div>
<div class="section" id="contact">
<h1>Contact<a class="headerlink" href="#contact" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p>Liguo Wang: wang.liguo AT mayo.edu</p></li>
<li><p>Wei Li: wei.li AT uci.edu</p></li>
</ul>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Release history</a></li>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#installation">Installation</a><ul>
<li><a class="reference internal" href="#prerequisite">Prerequisite</a></li>
<li><a class="reference internal" href="#install-cpat-using-pip3">install CPAT using pip3</a></li>
</ul>
</li>
<li><a class="reference internal" href="#run-cpat-online">Run CPAT online</a></li>
<li><a class="reference internal" href="#run-cpat-on-local-computer">Run CPAT on local computer</a><ul>
<li><a class="reference internal" href="#input-files">Input files</a></li>
<li><a class="reference internal" href="#command-line-options">Command line options</a></li>
<li><a class="reference internal" href="#examples">Examples</a></li>
<li><a class="reference internal" href="#output-files">output files</a></li>
</ul>
</li>
<li><a class="reference internal" href="#build-hexamer-table">Build hexamer table</a></li>
<li><a class="reference internal" href="#build-logit-model">Build Logit model</a></li>
<li><a class="reference internal" href="#use-cpat-to-detect-orf">Use CPAT to detect ORF</a><ul>
<li><a class="reference internal" href="#prepare-data">Prepare data</a></li>
<li><a class="reference internal" href="#run-cpat">Run CPAT</a></li>
<li><a class="reference internal" href="#check-the-results">Check the results</a></li>
</ul>
</li>
<li><a class="reference internal" href="#how-to-choose-cutoff">How to choose cutoff</a></li>
<li><a class="reference internal" href="#how-to-prepare-training-dataset">How to prepare training dataset</a></li>
<li><a class="reference internal" href="#evaluating-performance">Evaluating Performance</a></li>
<li><a class="reference internal" href="#comparison">Comparison</a></li>
<li><a class="reference internal" href="#license">LICENSE</a></li>
<li><a class="reference internal" href="#reference">Reference</a></li>
<li><a class="reference internal" href="#contact">Contact</a></li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="#">CPAT 3.0.0 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Release history</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, Liguo Wang.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 3.1.2.
    </div>
  </body>
</html>